### PR TITLE
Wrapping parser for Rust

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
This PR is a first part of work towards integrating parser written in Scala with components written in Rust. The design is:
* parser stays in Scala (if needed might be rewritten to Rust in future)
* paresr is called from Rust in the IDE (i.e. the browser side)
* Rust has its own datastructures for AST, paired with Scala ones
* Scala AST can be serialized and then deserialized in Rust code
* when working in browser (i.e. JS/Wasm world) parser will be compiled to JS using Scala.JS and called from Wasm using Wasm binding
* when used natively (e.g. for development, debugging, unit testing) parser can be set up as server and called using a simple websocket protocol

This PR consists of three parts:
1. Adding support for JSON serialization to AST in Scala. It uses `circe` — to enable its auto-deriving a number of code adjustments was necessary (mostly: traits must be sealed, case classes final and uniquely named, abstract classes are no-no). 
With the changes, AST can be serialized to JSON using `circe`'s `asJson` extension method.
1. Implementing a `ParserService` — a simple WebSocket server that wraps parser.
1. Implementing a `parser` create in Rust — that provides parser API for Rust and has two implementation: a websocket client when compiling natively and Wasm wrapper otherwise. Currently only a websocket client is implemented, Wasm wrapper is just a stub. 

Remaining work includes:
* exposing "real AST" and deserializing it from parser; (when #296 and #297 are done)
* wrapping JS parser for the web use case (when #257 is done)
* setting up CI support for the Rust parts

### Important Notes
Here is my first code written in Rust — let me know if I missed some idioms.
The parser websocket client/server are meant to be help in development, not "production" code.

### Checklist
Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [ ] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [ ] All code has been tested where possible.

### CODE REVIEW

File: common/scala/parser-service/src/main/scala/org/enso/ParserService.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+object ParserService {
+  val HOSTNAME_VAR = "ENSO_PARSER_HOSTNAME"
+  val PORT_VAR     = "ENSO_PARSER_PORT"
+
+  val DEFAULT_PORT     = 30615
+  val DEFAULT_HOSTNAME = "localhost"
+
+  def configFromEnv(): Server.Config = {
+    val hostname = sys.env.getOrElse(HOSTNAME_VAR, DEFAULT_HOSTNAME)
```
Comment (iamrecursion) [inline common/scala/parser-service/src/main/scala/org/enso/ParserService.scala]: I really dislike using env-based config rather than a configuration file. What do you think about it?

File: common/scala/parser-service/src/main/scala/org/enso/parserservice/Server.scala [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    case req @ HttpRequest(GET, _, _, _, _) =>
+      req.header[UpgradeToWebSocket] match {
+        case Some(upgrade) =>
+          println("Establishing a new connection")
+          upgrade.handleMessages(handlerFlow)
+        case None =>
+          HttpResponse(400, entity = "Not a valid websocket request!")
+      }
+    case r: HttpRequest =>
+      r.discardEntityBytes() // important to drain incoming HTTP Entity stream
```
Comment (iamrecursion) [inline common/scala/parser-service/src/main/scala/org/enso/parserservice/Server.scala]: Why? This kind of comment should be relegated to a source note. Same for the `//ignore binary ...` one above.

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala [hunk 3]
```diff
@@ -678,19 +692,20 @@ object AST {
       type Fmt      = ASTOf[FmtOf]
       type Unclosed = ASTOf[UnclosedOf]
 
-      case class LineOf[+T](off: Int, elem: List[T])
+      final case class LineOf[+T](off: Int, elem: List[T])
```
Comment (iamrecursion) [inline common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala]: Does this not introduce warnings about inability to check references to outer types at runtime?

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala [hunk 4]
```diff
@@ -1386,8 +1401,7 @@ object AST {
           : List[AST.Macro.Match.Segment] => List[AST.Macro.Match.Segment] =
           _.map(_.map({
             case m @ Pattern.Match.Nothing(_) => m
-            case m =>
-              unapplyValidChecker(unapplyFullChecker(m))
+            case m                            => unapplyValidChecker(unapplyFullChecker(m))
```
Comment (iamrecursion) [inline common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala]: No idea why the formatter has done this. It shouldn't overflow the line width.

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/meta/Pattern.scala [hunk 5]
```diff
@@ -116,12 +117,12 @@ object Pattern {
     Block(spaced) |
     Macro(spaced) |
     Invalid(spaced)
-  def Any(spaced: Boolean): Pattern = Any(Some(spaced))
-  def ErrTillEnd(msg: String)   = Any().tillEnd.err(msg)
-  def ErrUnmatched(msg: String) = End() | ErrTillEnd(msg)
-  def Expr()                    = Any().many1.build
-  def NonSpacedExpr()           = Any(spaced = false).many1.build
-  def NonSpacedExpr_()          = (Any().but(Block()) :: Any(spaced = false).many).build
+  def Any(spaced: Boolean): Pattern                = Any(Some(spaced))
```
Comment (iamrecursion) [inline common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/meta/Pattern.scala]: Is this formatted by scalafmt? It really, _really_ shouldn't be overflowing 80 like this.

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala [hunk 6]
```diff
@@ -21,18 +21,34 @@ object Escape {
   sealed trait Unicode extends Escape
   object Unicode {
 
-    final case class Invalid(unicode: Unicode) extends Unicode {
+    // NOTE [mwu]
```
Comment (iamrecursion) [inline common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala]: This should be a source note. Also that's a very ugly restriction. :c

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala [hunk 7]
```diff
@@ -77,66 +93,236 @@ object Escape {
     }
   }
 
-
-  abstract class Simple(val code: Int) extends Escape{
-    def name = toString
-    val repr = name
+  case object Slash extends Escape {
```
Comment (iamrecursion) [inline common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala]: What motivates all of these changes? It's far more verbose now.

File: .jvmopts [hunk 8]
```diff
@@ -1,2 +1,2 @@
 -Xss8M
--Xmx3072M
\ No newline at end of file
+-Xmx6144M
```
[the remark this answers is not in the document]
Comment (wdanilo) [reply .jvmopts]: I've got more on my computer, can we increase it to `32Gb` please? :D

File: common/rust/parser/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "parser"
+version = "0.1.0"
+authors = ["mwu <mwu-tow@gazeta.pl>"]
+edition = "2018"
+
+[lib]
+crate-type = ["cdylib", "rlib"]
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (wdanilo) [inline common/rust/parser/Cargo.toml]: Delete

File: common/rust/parser/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "parser"
+version = "0.1.0"
+authors = ["mwu <mwu-tow@gazeta.pl>"]
+edition = "2018"
+
+[lib]
+crate-type = ["cdylib", "rlib"]
```
Comment (wdanilo) [inline common/rust/parser/Cargo.toml]: Why `rlib` is needed here?

File: common/rust/parser/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
+wasm-bindgen = "0.2"
+wasm-bindgen-test = "0.2"
```
Comment (wdanilo) [inline common/rust/parser/Cargo.toml]: very minor - alignment

File: common/rust/parser/src/api.rs [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use Error::*;
+        match self {
+            ParsingError(msg) => write!(f, "Internal parser error: {:?}", msg),
+            InteropError(error) => {
+                write!(f, "Interop error: {}", error.to_string())
+            }
+        }
+    }
+}
```
Comment (wdanilo) [inline common/rust/parser/src/api.rs]: Define errors using utils exported from our prelude, see `backend/webgl/src/...` in BaseGL repo as an example:

```rust
#[derive(Debug, Fail, From)]
pub enum Error {
    #[fail(display = "Unable to create {}.", target)]
    Create { target: ErrorTarget },
    #[fail(display = "Unable to compile {}: {}", target, message)]
    Compile { target: ErrorTarget, message: String },
}
```

File: common/rust/parser/src/api.rs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+/////////
+// AST //
+/////////
+
+// TODO: placeholder until we have real AST, see: https://github.com/luna/enso/issues/296
+pub type AST = String;
+
+///////////
+// Error //
+///////////
```
Comment (wdanilo) [inline common/rust/parser/src/api.rs]: Very minor again, we use the following stylefor sections in Rust:

```
// =============
// === Error ===
// =============

File: common/rust/parser/src/jsclient.rs [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "JS parser client has not been yet implemented!")
+    }
+}
+
+pub struct Client {}
+
+impl Client {
+    #[allow(dead_code)] // avoid warnings when compiling natively and having this usage cfg-ed out
```
Comment (wdanilo) [inline common/rust/parser/src/jsclient.rs]: We should enable this conditionally then

File: common/rust/parser/src/jsclient.rs [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    pub fn new() -> Result<Client> {
+        Err(Error::NotImplemented)
+    }
+}
+
+impl Parser for Client {
+    fn parse(&mut self, _program: String) -> api::Result<api::AST> {
+        Err(api::Error::InteropError(Box::new(Error::NotImplemented)))
```
Comment (wdanilo) [inline common/rust/parser/src/jsclient.rs]: I would definitely create abstraction here. `Box::new` is implementation detail. Make a smart constructor for it instead.

File: common/rust/parser/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,20 @@
+pub mod api;
+
+mod jsclient;
+mod wsclient;
+
+type ParserWrapper = Box<dyn api::Parser>;
```
Comment (wdanilo) [inline common/rust/parser/src/lib.rs]: I don't like the naming here as it leaks the internal information. This type should be named `Parser` because it is just instance of Parser. While the interface should be renamed to `IsParser` instead.

File: common/rust/parser/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+
+#[cfg(not(target_arch = "wasm32"))]
+pub fn new_parser() -> api::Result<ParserWrapper> {
+    let client = wsclient::Client::new()?;
+    let parser = Box::new(client);
+    Ok(parser)
+}
```
Comment (wdanilo) [inline common/rust/parser/src/lib.rs]: The type above should be converted to newtype and these functions should be associated with the type, so you can write `Parser::new(...)` instead.

File: common/rust/parser/src/main.rs [hunk 18]
```diff
@@ -0,0 +1,17 @@
+/// Simple interactive tester - calls parser with its argument (or a
+/// hardcoded default) and prints the result
+fn main() {
+    let default_input = String::from("import Foo.Bar\nfoo = a + 2");
+    let input = std::env::args().skip(1).next().unwrap_or(default_input);
+    println!("Will parse: {}", input);
+
+    let mut parser = parser::new_parser()
+        .unwrap_or_else(|e| panic!("Failed to create a parser: {:?}", e));
```
Comment (wdanilo) [inline common/rust/parser/src/main.rs]: This is sometimes very useful, I'd name it `Parser::new_or_panic(...)`.

File: common/rust/parser/src/wsclient.rs [hunk 19]
```diff
@@ -0,0 +1,201 @@
+#![cfg(not(target_arch = "wasm32"))]
+
+use websocket::{
+    stream::sync::TcpStream, ClientBuilder, Message, OwnedMessage,
+};
+
+type WsTcpClient = websocket::sync::Client<TcpStream>;
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: defining type between `use` things is strange to me :D

File: common/rust/parser/src/wsclient.rs [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+//////////////////////////
+// Constants & literals //
+//////////////////////////
+
+const LOCALHOST: &str = "localhost";
+const DEFAULT_PORT: i32 = 30615;
+const DEFAULT_HOSTNAME: &str = LOCALHOST;
+
+const HOSTNAME_VAR: &str = "ENSO_PARSER_HOSTNAME";
+const PORT_VAR: &str = "ENSO_PARSER_PORT";
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: alignment

File: common/rust/parser/src/wsclient.rs [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub enum Error {
+    WrongUrl(websocket::client::ParseError),
+    ConnectivityError(websocket::WebSocketError),
+    NonTextResponse(websocket::OwnedMessage),
+    JsonSerializationError(serde_json::error::Error),
+}
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: Alignment - we align the arguments vertically

File: common/rust/parser/src/wsclient.rs [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub enum Request {
+    ParseRequest { program: String },
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub enum Response {
+    Success { ast: String },
+    Error { message: String },
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: alignment of `{`, `:`, and `}`

File: common/rust/parser/src/wsclient.rs [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+impl Config {
+    pub fn address_string(&self) -> String {
+        format!("ws://{}:{}", self.host, self.port)
+    }
+
+    pub fn from_env() -> Config {
+        let host = env_var_or(HOSTNAME_VAR, DEFAULT_HOSTNAME);
+        let port = env_var_or(PORT_VAR, "").parse().unwrap_or(DEFAULT_PORT);
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: instead of `""` use `default()`.

File: common/rust/parser/src/wsclient.rs [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+        Config { host, port }
+    }
+}
+
+fn env_var_or(varname: &str, default_value: &str) -> String {
+    std::env::var(varname).unwrap_or_else(|_| String::from(default_value))
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: `x.into()` is shorter than `x.to_string()` which is shorter than `String::from(x)`.

File: common/rust/parser/src/wsclient.rs [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+
+pub struct Client {
+    connection: WsTcpClient,
+}
+
+impl Client {
+    pub fn from_conf(config: &Config) -> Result<Client> {
+        let address = config.address_string();
+        let mut builder = ClientBuilder::new(&address)?;
+        let connection = builder.connect_insecure()?;
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: alignment of `=`

File: common/rust/parser/src/wsclient.rs [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+        Ok(Client { connection })
+    }
+
+    pub fn new() -> Result<Client> {
+        let config = Config::from_env();
+        println!("Connecting to {}", config.address_string());
+        let client = Client::from_conf(&config)?;
+        println!("Established connection with {}", config.address_string());
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: Closer to merge, lets move to our logging framework here!  Basegl has some amazing hierarchical logging utils for web-console.

File: common/rust/parser/src/wsclient.rs [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+        self.connection.send_message(&message)?;
+        Ok(())
+    }
+
+    fn recv_response(&mut self) -> Result<Response> {
+        let response = self.connection.recv_message()?;
+        match response {
+            OwnedMessage::Text(text) => {
+                let response = serde_json::from_str(&text)?;
+                Ok(response)
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: Why not `Ok(serde_json::from_str(&text)?)`? I would understand if the var name was more descriptive, but it actually aliases with other var and I dont fully understand its meaning here :D

File: common/rust/parser/src/wsclient.rs [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+    }
+}
+
+impl api::Parser for Client {
+    fn parse(&mut self, program: String) -> api::Result<api::AST> {
+        let request = Request::ParseRequest { program };
+        let response = self.rpc_call(request)?;
+        match response {
+            Response::Success { ast } => Ok(ast),
+            Response::Error { message } => Err(ParsingError(message)),
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: alignment of `{`, `}`, and `=>`

File: common/rust/parser/src/wsclient.rs [hunk 29]
```diff
@@ -0,0 +1,201 @@
...
+#[test]
+fn wrong_url_reported() {
+    let invalid_hostname = String::from("bgjhkb 7");
+    let wrong_config = Config {
+        host: invalid_hostname,
+        port: 8080,
+    };
+    let client = Client::from_conf(&wrong_config);
+
+    if let Err(WrongUrl(_)) = client {
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: Oh god, please, can we not use this syntax? :D Maybe its just me, but Im always reading it 10 times until I get what this line really does :D

File: common/rust/parser/src/wsclient.rs [hunk 30]
```diff
@@ -0,0 +1,201 @@
...
+    } else {
+        assert!(false, "expected WrongUrl error");
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: This is very minor, but this is error prone for refctoring. The `WringUrl` name should be get using reflection utils

File: common/scala/parser-service/src/main/scala/org/enso/ParserService.scala (hunk 1 above)
Comment (wdanilo) [reply common/scala/parser-service/src/main/scala/org/enso/ParserService.scala]: Why? Env-based config + defaults is much superior than file-based config. You can just run it and it works. With files you alwaysneeds to create one, which is very against 0-config philosophy. In the best scenario - there should be 0-config approach + config file + env vars overriding these settings, but as far as I understand, this is only a debugging util, so I'm happy without config files now.

File: common/scala/parser-service/src/main/scala/org/enso/ParserService.scala [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+
+case class ParserService(config: Server.Config) extends Server with Protocol {
+  import parserservice._
+  import Protocol._
+
+  val parser = new Parser()
+  def serializeAst(ast: AST.Module): String =
+    ast.asJson.noSpaces
+
+  def runParser(program: String): AST = new Parser().run(new Reader(program))
```
Comment (wdanilo) [inline common/scala/parser-service/src/main/scala/org/enso/ParserService.scala]: Very minor, I would do newline after `=` to increase redability

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala (hunk 3 above)
Comment (wdanilo) [reply common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala]: Good point, I would be curious as well. Btw, I hate this thing in Scala.

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala (hunk 4 above)
Comment (wdanilo) [reply common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala]: Yeah, it happens to me sometimes as well. In such case I just put a blank line between the cases.

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/meta/Pattern.scala [hunk 32]
```diff
@@ -80,28 +81,28 @@ object Pattern {
...
   object Text {
-    def apply():                Text = Text(None)
+    def apply(): Text                = Text(None)
     def apply(spaced: Boolean): Text = Text(Some(spaced))
   }
   object Block {
-    def apply():                Block = Block(None)
+    def apply(): Block                = Block(None)
     def apply(spaced: Boolean): Block = Block(Some(spaced))
   }
```
Comment (wdanilo) [inline common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/meta/Pattern.scala]: We definitely have some strange behavior here. I see this change nth time. Sometimes it is getting on right, sometimes on left. Just a note to check why it happens.

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala (hunk 6 above)
Comment (wdanilo) [reply common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala]: @mwu-tow we can create here alias `type Invalid = InvalidUnicode` and `val Invalid = InvalidUnicode` right?

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala (hunk 7 above)
Comment (wdanilo) [reply common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala]: Shapeless does not work if:
- you use abstract classes
- you use not sealed traits
- you use not final case classes
- your file is not ordered properly (ordering of defs has a meaning -.-)
- and many others

This implies that deriving json does not work either

File: common/rust/parser/src/wsclient.rs (hunk 23 above)
Comment (mwu-tow) [reply common/rust/parser/src/wsclient.rs]: done

File: common/rust/parser/src/wsclient.rs (hunk 24 above)
Comment (mwu-tow) [reply common/rust/parser/src/wsclient.rs]: done

File: common/rust/parser/src/wsclient.rs (hunk 25 above)
Comment (mwu-tow) [reply common/rust/parser/src/wsclient.rs]: done

File: common/rust/parser/src/wsclient.rs (hunk 26 above)
Comment (mwu-tow) [reply common/rust/parser/src/wsclient.rs]: as soon as we have common logging framework on this repo

File: common/rust/parser/src/wsclient.rs (hunk 27 above)
Comment (mwu-tow) [reply common/rust/parser/src/wsclient.rs]: done, good point

File: common/rust/parser/src/wsclient.rs (hunk 28 above)
Comment (mwu-tow) [reply common/rust/parser/src/wsclient.rs]: done.

File: common/rust/parser/src/wsclient.rs (hunk 30 above)
Comment (mwu-tow) [reply common/rust/parser/src/wsclient.rs]: what is the most idiomatic way to achieve this?

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala (hunk 7 above)
Comment (mwu-tow) [reply common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala]: I am not sure if abstract classes truly creak this or if other issues were causing this to look as if.
Might reinvestigate later.

File: common/scala/syntax/definition/src/main/scala/org/enso/data/Tree.scala [hunk 33]
```diff
@@ -34,4 +37,23 @@ object Tree {
   def apply[K, V](): Tree[K, V] = new Tree(None, Map())
   def apply[K, V](deps: (List[K], V)*): Tree[K, V] =
     deps.foldLeft(Tree[K, V]())(_ + _)
+
+  //////////////////
+  // JSON support //
+  //////////////////
+
+  // Note [MWU]
```
[the remark this answers is not in the document]
Comment (mwu-tow) [reply common/scala/syntax/definition/src/main/scala/org/enso/data/Tree.scala]: done

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala (hunk 6 above)
Comment (mwu-tow) [reply common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala]: Yes, done.

File: common/rust/parser/src/main.rs (hunk 18 above)
Comment (mwu-tow) [reply common/rust/parser/src/main.rs]: done

File: common/rust/parser/src/wsclient.rs (hunk 19 above)
Comment (mwu-tow) [reply common/rust/parser/src/wsclient.rs]: done

**Comment by mwu-tow** [comment]
I've pushed an update that applies most of requested changes. Haven't looked into scalafmt going wild and abstract classes reinvestigation. (and not sure if I should, given timelines).
Will also likely clean up the testing code in the Main.

@wdanilo As for formatting — I wasted quite a much time, applying comments (not all the "done"s)… then rustfmt broke it all again. We need to fix the tooling first.

@iamrecursion I've tried adding documentation for relevant code parts where I had anything to add besides repeating identifiers. Please let me know if there are any parts left that would benefit from more documentation.

As for tests — for now that's basically "code compiles". If this compiles, then it means that we can call asJson on AST — and that circe is able to derive all the encoders. Verifying that it serializes to something useful will be done when connecting the Rust implementation of AST.
Basically then we will adjust the serialization schema here to conform with the Rust one. It wouldn't make sense to test it now, as it will change.

File: Cargo.lock [hunk 34]
```diff
@@ -0,0 +1,1265 @@
+# This file is automatically @generated by Cargo.
```
[the remark this answers is not in the document]
Comment (iamrecursion) [reply Cargo.lock]: Yeah, that's a very good point. I do think removing it for now is probably the best bet, though. We can always add it back later.

File: common/scala/parser-service/src/main/scala/org/enso/ParserService.scala (hunk 1 above)
Comment (iamrecursion) [reply common/scala/parser-service/src/main/scala/org/enso/ParserService.scala]: Env-based config is more obtuse in situations where you _want_ things to be configurable, rather than having configuration as an afterthought. I'm not going to die on this hill, though, so this isn't a review show-stopper.

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala (hunk 3 above)
Comment (iamrecursion) [reply common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala]: As far as I'm aware it affects _all_ final case classes, but who knows.

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala (hunk 7 above)
Comment (iamrecursion) [reply common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala]: Only if it doesn't take too much time. I can accept changes needed for the sake of other libraries, I just don't like it!

File: common/rust/parser/src/api.rs [hunk 35]
```diff
@@ -0,0 +1,42 @@
...
+pub trait IsParser {
+    fn parse(&mut self, program: String) -> Result<AST>;
+}
+
+// =========
+// == AST ==
+// =========
+
+// TODO: placeholder until we have real AST, see:
+// https:==github.com=luna=enso=issues=296
```
Comment (iamrecursion) [inline common/rust/parser/src/api.rs]: Find and replace issue here?

Comment (mwu-tow) [reply common/rust/parser/src/api.rs]: Yes, thank you for catching this.

File: common/rust/parser/Cargo.toml (hunk 10 above)
Comment (farmaazon) [reply common/rust/parser/Cargo.toml]: Interesting, here I specified https://github.com/luna/basegl/blob/wip/ao/msdfgen-integration/lib/core/msdf-sys/Cargo.toml and `wasm-pack test` works perfectly, at least on linux...

File: common/rust/parser/src/jsclient.rs (hunk 14 above)
Comment (wdanilo) [reply common/rust/parser/src/jsclient.rs]: I don't get it, lets talk via phone about it. Even if we leave it as is, the line is >80 chars :P

File: common/rust/parser/src/lib.rs [hunk 36]
```diff
@@ -0,0 +1,41 @@
...
+    /// Obtains a default parser implementation, panicking in case of failure.
+    pub fn new_or_panic() -> Parser {
+        Parser::new()
+            .unwrap_or_else(|e| panic!("Failed to create a parser: {:?}", e))
+    }
+}
+
+impl api::IsParser for Parser {
+    fn parse(&mut self, program: String) -> api::Result<api::AST> {
+        self.0.parse(program)
```
Comment (wdanilo) [inline common/rust/parser/src/lib.rs]: Let's use Shrinkwrap instead!

File: common/rust/parser/src/wsclient.rs [hunk 37]
```diff
@@ -0,0 +1,227 @@
...
+// ==========================
+// == Constants & literals ==
+// ==========================
+
+pub const LOCALHOST: &str = "localhost";
+pub const DEFAULT_PORT: i32 = 30615;
+pub const DEFAULT_HOSTNAME: &str = LOCALHOST;
+
+pub const HOSTNAME_VAR: &str = "ENSO_PARSER_HOSTNAME";
+pub const PORT_VAR: &str = "ENSO_PARSER_PORT";
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: alignment

File: common/rust/parser/src/wsclient.rs (hunk 21 above)
Comment (wdanilo) [reply common/rust/parser/src/wsclient.rs]: Hmm, I don't see it. Have you committed the change?

File: common/rust/parser/src/wsclient.rs [hunk 38]
```diff
@@ -0,0 +1,227 @@
...
+                    write!(f, "Peer closed connection: {:?}", close_data)
+                }
+                _ => write!(f, "Expected text response, got: {:?}", msg),
+            },
+            JsonSerializationError(msg) => {
+                write!(f, "JSON (de)serialization failed: {:?}", msg)
+            }
+        }
+    }
+}
```
Comment (wdanilo) [inline common/rust/parser/src/wsclient.rs]: According to previous comment - let's use the Prelude exposed libraries for this.

File: common/rust/parser/src/wsclient.rs (hunk 30 above)
Comment (wdanilo) [reply common/rust/parser/src/wsclient.rs]: https://doc.rust-lang.org/std/any/fn.type_name.html

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala (hunk 7 above)
Comment (mwu-tow) [reply common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/ast/text/Escape.scala]: I did some experiments but wasn't able to solve the issue in reasonable time. I believe it *should* be possible… and abstract classes seem to work with circe on small examples. However here… when I try introducing it, I get errors that I can't efficiently reduce because build times explode (7-8 min before error). 
Unfortunately we'll to leave this as-presented for now, as we don't have more time to spare on this. Sorry. :(
